Move all user-facing text to a central location

somebody1234 opened the pull request:
### Pull Request Description
- Closes https://github.com/enso-org/cloud-v2/issues/861
  - Adds a `getText` function and React Context to abstract away all text that is shown to users

The main immediate benefit is making all text much more discoverable - both being able to know of every piece of text used across the entire application, and making it a lot easier to refactor certain strings when needed rather than having to hunt for strings to replace (and potentially miss one).

The longer term benefit is that this will make it easy to add localization, by simply adding another JSON file with the same keys as the existing one, and adding a little bit of logic.

### Important Notes
None

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
WIP: Add support for localization

somebody1234 committed:
Reformat to 2 spaces

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Finish replacing all texts with localizable texts

somebody1234 committed:
Also make shortcut names localizable

somebody1234 committed:
Fix type errors; fix lint errors

somebody1234 committed:
Move user-facing auth messages to `english.json`

somebody1234 committed:
WIP: Move user-facing `remoteBackend.ts` messages to `english.json`

somebody1234 committed:
Finish refactoring `remoteBackend.ts` to centralized text

somebody1234 committed:
Fix minor errors

somebody1234 committed:
Refactor `toastAndLog` to accept a text id instead of a string

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Trigger CI

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Fix type errors

somebody1234 committed:
Fix other lint errors

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Fix type errors

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Fix new lint errors caused by the new lint

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Fix type errors

indiv0 commented on app/ide-desktop/lib/dashboard/src/components/MenuEntry.tsx:
```diff
@@ -68,7 +70,7 @@ export default function MenuEntry(props: MenuEntryProps) {
       <div className="flex items-center gap-menu-entry whitespace-nowrap">
         <SvgMask src={info.icon ?? BlankIcon} color={info.color} className="size-icon" />
-        {label ?? info.name}
+        {label ?? getText(`${action}Shortcut`)}
```
This might be hard to find in refactoring later. Perhaps `action` should define a `shortcut` field with contents like `'FooActionShortcut'` and we reference that directly here like <code>getText(`${action.shortcut}`)</code>?

Similar to the TailwindCSS principle of not constructing class names by concatenating text, but rather by matching on possibilities. That is, `if item.size == small { "font-sm" } else { "font-lg" }` rather than `"font-" + item.size`.

Just a suggestion.

indiv0 commented on app/ide-desktop/lib/dashboard/src/components/dashboard/PermissionSelector.tsx:
```diff
@@ -195,7 +198,7 @@ export default function PermissionSelector(props: PermissionSelectorProps) {
           onClick={doShowPermissionTypeSelector}
         >
-          {permission.type}
+          {getText(`${permission.type}PermissionType`)}
```
ibid. re: concatenation

Applies to all other occurrences in this PR.

indiv0 commented on app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx:
```diff
@@ -184,24 +186,13 @@ export default function ProjectNameColumn(props: ProjectNameColumnProps) {
-                null
-              )
+              const projectId = backendModule.ProjectId(id)
+              const listedProject = await backend.getProjectDetails(projectId, '(unknown)')
```
Why this string instead of `null`? Seems like `null` is the intended value that we should pass in a case like this, rather than passing a string then potentially matching on it later.

somebody1234 replied:
the `title` is for logging purposes only - i figured it's better to require the title in almost every case, as there are only one or two situations where we don't actually know the title. for this specific case, i've switched it to use `file.name` for now.

indiv0 commented on app/ide-desktop/lib/dashboard/src/layouts/Samples.tsx:
```diff
@@ -35,6 +37,8 @@ const DUMMY_LIKE_COUNT = 10
 /** Template metadata. */
 export interface Sample {
   readonly title: string
+  /** These should ideally be localized, however, as this is planned to be UGC, it is unlikely
+   * that this will be feasible. */
```
Define acronyms prior to usage, or better yet don't use acronyms.

indiv0 reviewed (approved):
Aside from minor comments the code looks good to me.

QA passed ✅ 

That being said, not sure if I agree with the rationale for this PR. It seems like moving error message declarations away from where they're used adds an additional step when trying to diagnose what caused an error. Also, [localization is usually much trickier](https://metacpan.org/dist/Locale-Maketext/view/lib/Locale/Maketext/TPJ13.pod#A_Localization_Horror_Story:_It_Could_Happen_To_You) than it appears at first glance. I'm not saying it's not a problem worth solving, but it seems like the kind of task that we should tackle once we have a demonstrable (i.e., customer-requested) need for it. Efforts to solve localization prior to involving actual translators for specific languages are likely to be duplicate work, since a simple `getText` will need to be replaced anyways in order to work correctly.

somebody1234 commented:
> adds an additional step when trying to diagnose what caused an error

true, but not sure how much this matters in practice - the actual cause should show up somewhere in the stacktrace anyway.
(on that note - perhaps we should preserve sourcemaps in the built app? i've found that to cause a decent amount of trouble when debugging errors there...)

> since a simple `getText` will need to be replaced anyways in order to work correctly.

afaict this is unlikely: the API is designed as an abstraction - all you pass in are the key and the parameters (if any), and while the current implementation *is* just a simple text replacement, there's nothing saying it can't be replaced with a more complex system later.

(if i missed something please do correct me though.)

somebody1234 commented:
also note (as noted in the description) that another advantage (perhaps the main one) is collecting all the text in one place, so that anyone who wants to review or adjust the text, can be sure that they haven't missed some text.

@jdunkerley do you have any thoughts on this? would this be useful at all?

just to clarify for everyone - i don't have any issues with shelving this PR indefinitely, especially as there is no short term benefit - it doesn't add a new feature, and it doesn't make the codebase easier to maintain.

indiv0 commented:
> true, but not sure how much this matters in practice - the actual cause should show up somewhere in the stacktrace anyway. (on that note - perhaps we should preserve sourcemaps in the built app? i've found that to cause a decent amount of trouble when debugging errors there...)

👍 for source maps. I'm of the opinion that they should be included for all builds. Not sure if there's any reason why we don't currently include them. And yeah, I suppose they would resolve the issue of error diagnostics via error messages.

> afaict this is unlikely: the API is designed as an abstraction - all you pass in are the key and the parameters (if any), and while the current implementation is just a simple text replacement, there's nothing saying it can't be replaced with a more complex system later.

That's kind of my point, it *will* need to be replaced with a more complex system later. The current implementation won't be sufficient for localization. For example, when passing in parameters we'll need to know if they refer to a quantity, or a noun, etc. So the value for the `getText` implementation *in its current state* is just centralizing error messages.

> also note (as noted in the description) that another advantage (perhaps the main one) is collecting all the text in one place, so that anyone who wants to review or adjust the text, can be sure that they haven't missed some text.

This *does* provide value, as you mentioned -- it makes changing messages easier. I just wanted to be sure we weighed that against the downside of moving said messages away from where they're used in the code.

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Address review

indiv0 reviewed (approved):
LGTM

PabloBuchu commented:
Bit anecdotal evidence but most of my projects had user facing texts in single file ;)) imho it's another convention but it helps keep texts more or less consistent especially when you have more developers working on the same project. It is true its a bit easier also to add localisation which we probably work on sooner than later

| It seems like moving error message declarations away from where they're used adds an additional step when trying to diagnose what caused an error.

I disagree here. We added much more complexity than that which makes development and debugging bit harder.

somebody1234 commented:
Closing - it's been decided that it doesn't bring enough benefit to merit such invasive changes

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 commented:
reopened after addressing concerns:
- might forget to add new messages: this should be covered by the new lint added in this PR.
- might make us a little slower: possibly - i *think* it shouldn't hurt for the most part, since instead of literal text it's one (auto-)import per component plus an identifier, plus one line added to a JSON file. but i guess only time will tell whether this *really* slows us down or not.

MrFlashAccount reviewed (changes requested):
Why we're creating a custom solution instead of using a standard one? (i18n is pretty solved problem in the js ecosystem already).

somebody1234 committed:
Fix type error

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Simplify `DuplicateAssetsModal.tsx`

somebody1234 committed:
Change some `string` parameters for `getText` to `number` as appropriate

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text

somebody1234 committed:
Fix type error

indiv0 commented:
> Why we're creating a custom solution instead of using a standard one? (i18n is pretty solved problem in the js ecosystem already).

Because we don't need localization yet. This PR is just for moving the text out to a central location. Proper l18n is a separate issue, for some point in the future, and should indeed be handled with an appropriate library.

somebody1234 committed:
Merge branch 'develop' into wip/sb/centralize-text
